Improve NetAtmo sensors update logic

glpatcern opened the pull request:
## Description:

Following PR #14546, a suggestion was made to synchronize the data polling to HA with the effective refresh in the NetAtmo cloud.

This PR changes the throttling logic of the `update` method such that data is refreshed every 10 minutes and at the best time to fetch the fresh data from the cloud.

Note that following code review, the `Throttle` decorator is not used, but its spirit is there, along with a lock to protect the execution of the remote update.

I believe the same approach could be used in other sensors, whenever the timing would represent an added value (this was the case for me to have a quicker reaction in some automations).

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable):
```
sensor:
  - platform: netatmo
    modules:
      <your_module>:
        - ...
        - lastupdated
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

glpatcern committed:
Added a "last update" sensor that could be used by automations + cosmetic changes

glpatcern committed:
Improved the update logic of sensor data

The platform is now continuously adjusting the refresh interval
in order to synchronize with the expected next update from the
NetAtmo cloud. This significantly improves reaction time of
automations while keeping the refresh time to the recommended
value (10 minutes).

glpatcern committed:
Linting

balloob commented:
Just don't pass `True` as second argument to add_devices and Home Assistant won't call update when devices are added to Home Assistant.

glpatcern committed:
Incorporated the advanced Throttle class to support adaptive
throttling, as opposed to integrating it in the core framework.

Following code review, it was suggested to implement the
specialised Throttle class in this platform instead of making a
change in the general util package. Except that the required change
(about 4 LoC) is part of the only relevant piece of code of that
class, therefore this commit includes a full copy of the Throttle
class from homeassistant.util, plus the extra feature to support
adaptive throttling.

glpatcern commented:
As discussed in #14865, what I propose here cannot be implemented by what Home Assistant provides.

The last commit makes this PR working independently from changes in the core framework. However, it introduces a lot of duplicated code.

So I'd like some feedback from the community: if you believe that adaptive throttling to "follow the cloud" is a plus for NetAtmo and possibly other sensors, this PR can go ahead - and if there's enough motivation, my advice as a developer would be to merge back the last commit in the core framework, removing the duplicated code. Otherwise, I'm fine keeping this as a custom component for myself (it Works For Me very well) and close the PR.

glpatcern commented:
Removing the WIP tag as not needed any longer.
And "pinging" @heinemml @syssi @Danielhiversen, contributors in the discussion in #14546: any opinion from you about this adaptive polling?

balloob commented on homeassistant/components/sensor/netatmo.py:
```diff
@@ -313,3 +321,131 @@ def update(self):
+                newinterval = NETATMO_UPDATE_INTERVAL / 2
+            _LOGGER.warning("NetAtmo refresh interval reset to %d seconds",
+                            newinterval)
+            self.update = Throttle(timedelta(seconds=newinterval),
```
I think that it is getting overly complicated with the throttle.

Why not just keep track of a timestamp and that's it?

```python
if self._call_in_progress or dt_util.utcnow() < this._next_call:
    return

self._call_in_progress = True
try:
    self.next_call = newinterval
    # All the other fechting stuff
finally:
    self._call_in_progress = False
```

glpatcern replied:
Surely the throttle has its intrinsic elegance ;-) but yes, this is a good solution as well, and it avoids the code duplication. I'll look into that.

glpatcern committed:
Cosmetic changes on the introduced "last updated" sensor

glpatcern committed:
Alternate implementation for the adaptive throttling

Ensure the updates from the cloud are throttled and adapted to the
last update time provided by NetAtmo, without using the Throttle
decorator. Similar logic and similar usage of a lock to protect
the execution of the remote update.

glpatcern committed:
Linting

glpatcern commented:
Great. Shall I create a PR in https://home-assistant.github.io to describe how to configure this (as per the `configuration.yaml` snippet in the description)?

balloob commented on homeassistant/components/sensor/netatmo.py:
```diff
@@ -285,6 +286,8 @@ def update(self):
                 self._state = "High"
             elif data['wifi_status'] <= 55:
                 self._state = "Full"
+        elif self.type == 'lastupdated':
```
Ugh, I missed this. We should not allow this. We never are allowed to store relative times in the state machine. Opening a PR to remove this feature.

balloob replied:
#15560
